drivers/led_strip/ws2812_uart.c

```diff
@@ -0,0 +1,342 @@
+ * The driver reuses the UART's hardware-generated start and stop bits as part
+ * of the on-wire symbol.
+ *  - The symbol's MSB ('1') maps to the inverted UART start bit.
+ *  - The inner bits are packed into the UART data payload.
+ *  - The symbol's LSB ('0') maps to the inverted UART stop bit.
+ *
+ * Configuration Constraint:
+ * This packing scheme imposes a constraint: the total number of bits in a
+ * UART frame (1 start + N data + 1 stop) must be an integer multiple of the
+ * symbol's length (`bits-per-symbol`). For example, if `data-bits` is set to 7,
+ * the 9-bit total frame size (1 + 7 + 1) is compatible with a `bits-per-symbol`
+ * of 3.
+ *
+ * Example: The WS2812 data stream `101` sent as symbols `110`, `100`, `110`
+ * and packed into one 9-bit UART frame (1 start + 7 data + 1 stop):
+ *                       d0 d1 d2 d3 d4 d5 d6
+ *                    __ __    __       __ __
+ *                ___|     |__|  |__ __|     |__ ...
+ *   Start Bit (high) ^                       ^ Stop Bit (low)
+ */
+
+#define DT_DRV_COMPAT worldsemi_ws2812_uart
+
+#include <string.h>
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/led_strip.h>
+#include <zephyr/drivers/uart.h>
+#include <zephyr/dt-bindings/led/led.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util.h>
+
+#define LOG_LEVEL CONFIG_LED_STRIP_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ws2812_uart);
+
+/* Each color channel is represented by 8 bits. */
+#define BITS_PER_COLOR_CHANNEL 8
+
+/*
+ * Helper macros to get UART frame configuration from the referenced UART's devicetree node.
+ */
+#define _DT_UART_NODE(inst)          DT_INST_PHANDLE(inst, uart)
+#define _DT_UART_DATA_BITS(inst)     DT_PROP_OR(_DT_UART_NODE(inst), data_bits, 8)
+/* Only UART_CFG_STOP_BITS_1 or 1 is supported. Other values will fail the BUILD_ASSERT below. */
+#define _DT_UART_STOP_BITS(inst)     DT_ENUM_IDX_OR(_DT_UART_NODE(inst), stop_bits, 1)
+#define _DT_UART_HAS_PARITY(inst)    (DT_ENUM_IDX(_DT_UART_NODE(inst), parity) != \
+				      UART_CFG_PARITY_NONE)
+#define _DT_UART_HAS_TX_INVERT(inst) (DT_PROP_OR(_DT_UART_NODE(inst), tx_invert, 0) != 0)
+
+/* The total number of bits for one UART frame transmission (start + data + parity + stop). */
+#define UART_FRAME_BITS_FROM_DT(inst) \
+	(1 + _DT_UART_DATA_BITS(inst) + _DT_UART_HAS_PARITY(inst) + _DT_UART_STOP_BITS(inst))
+
+struct ws2812_uart_cfg {
+	const struct device *uart_dev;
+	uint8_t *px_buf;
+	uint16_t one_symbol;
+	uint16_t zero_symbol;
+	uint8_t bits_per_symbol;
+	uint8_t num_colors;
+	const uint8_t *color_mapping;
+	size_t length;
+	uint16_t reset_delay;
+	uint8_t uart_frame_bits;
+};
+
+struct ws2812_uart_data {
+	struct k_mutex lock;
+	struct k_sem tx_done_sem;
+};
+
+static inline const struct ws2812_uart_cfg *dev_cfg(const struct device *dev)
+{
+	return dev->config;
+}
+
+static inline struct ws2812_uart_data *dev_data(const struct device *dev)
+{
+	return dev->data;
+}
```

Comment:
drop, just use dev->data and dev->config directly, this does not really make the code any more compact (you don't even need a casting) and it's not normally done for Zephyr drivers